spec/validators/email_mx_validator_spec.rb

```diff
@@ -58,65 +46,69 @@
-      allow(resolver).to receive(:getresources).with('example.com', Resolv::DNS::Resource::IN::A).and_return([])
-      allow(resolver).to receive(:getresources).with('example.com', Resolv::DNS::Resource::IN::AAAA).and_return([])
-      allow(resolver).to receive(:getresources).with('mail.example.com', Resolv::DNS::Resource::IN::A).and_return([])
-      allow(resolver).to receive(:getresources).with('mail.example.com', Resolv::DNS::Resource::IN::AAAA).and_return([])
-      allow(resolver).to receive(:timeouts=).and_return(nil)
-      allow(Resolv::DNS).to receive(:open).and_yield(resolver)
+      configure_resolver('example.com', mx: resolv_double_mx('mail.example.com'))
+      configure_resolver('mail.example.com')
 
       subject.validate(user)
       expect(user.errors).to have_received(:add)
     end
 
     it 'adds an error if the MX record is blacklisted' do
       EmailDomainBlock.create!(domain: 'mail.example.com')
-      resolver = instance_double(Resolv::DNS)
-
-      allow(resolver).to receive(:getresources)
-        .with('example.com', Resolv::DNS::Resource::IN::MX)
-        .and_return([instance_double(Resolv::DNS::Resource::MX, exchange: 'mail.example.com')])
-      allow(resolver).to receive(:getresources).with('example.com', Resolv::DNS::Resource::IN::A).and_return([])
-      allow(resolver).to receive(:getresources).with('example.com', Resolv::DNS::Resource::IN::AAAA).and_return([])
-      allow(resolver).to receive(:getresources).with('mail.example.com', Resolv::DNS::Resource::IN::A).and_return([instance_double(Resolv::DNS::Resource::IN::A, address: '2.3.4.5')])
-      allow(resolver).to receive(:getresources).with('mail.example.com', Resolv::DNS::Resource::IN::AAAA).and_return([instance_double(Resolv::DNS::Resource::IN::AAAA, address: 'fd00::2')])
-      allow(resolver).to receive(:timeouts=).and_return(nil)
-      allow(Resolv::DNS).to receive(:open).and_yield(resolver)
+
+      configure_resolver('example.com', mx: resolv_double_mx('mail.example.com'))
+      configure_resolver('mail.example.com')
 
       subject.validate(user)
       expect(user.errors).to have_received(:add)
     end
   end
+
+  def configure_resolver(domain, options = {})
+    allow(resolv_dns_double)
+      .to receive(:getresources)
+      .with(domain, Resolv::DNS::Resource::IN::MX)
+      .and_return(Array(options[:mx]))
+    allow(resolv_dns_double)
+      .to receive(:getresources)
+      .with(domain, Resolv::DNS::Resource::IN::A)
+      .and_return(Array(options[:a]))
+    allow(resolv_dns_double)
+      .to receive(:getresources)
+      .with(domain, Resolv::DNS::Resource::IN::AAAA)
+      .and_return(Array(options[:aaaa]))
+    allow(resolv_dns_double)
+      .to receive(:timeouts=)
+      .and_return(nil)
+    allow(Resolv::DNS)
+      .to receive(:open)
+      .and_yield(resolv_dns_double)
+  end
+
+  def resolv_dns_double
+    @resolv_dns_double ||= instance_double(Resolv::DNS)
+  end
```

Comment:
I find this pattern a little surprising, especially in a spec. Any reason for this not to be a local variable in `configure_resolver`?

Reply:
Oh yeah, that makes sense. I think I would prefer a `let(:resolv_dns_double) { instance_double(Resolv::DNS) }` then, but I won't block the PR if you think your current approach is better.

Reply:
Fine either way. Moved to a `let`.